perf: improve applyOrdering by avoid call to GetByCurrentId

chlunde opened the pull request:
This shaves of 14 seconds (one third) of the execution time for a kustomization tree with 4000 documents, from 40.68s to 27.41s

            0     0%  5.44%     18.42s 40.56%  sigs.k8s.io/kustomize/api/krusty.(*Kustomizer).applySortOrder
            0     0%  5.44%     18.40s 40.52%  sigs.k8s.io/kustomize/api/internal/builtins.applyOrdering
before

    (pprof) top20 -cum
    Showing nodes accounting for 5.85s, 12.88% of 45.41s total
    Dropped 622 nodes (cum <= 0.23s)
    Showing top 20 nodes out of 157
        flat  flat%   sum%        cum   cum%
            0     0%     0%     40.68s 89.58%  github.com/spf13/cobra.(*Command).Execute
            0     0%     0%     40.68s 89.58%  github.com/spf13/cobra.(*Command).ExecuteC
            0     0%     0%     40.68s 89.58%  github.com/spf13/cobra.(*Command).execute
            0     0%     0%     40.68s 89.58%  main.main
            0     0%     0%     40.68s 89.58%  runtime.main
            0     0%     0%     40.68s 89.58%  sigs.k8s.io/kustomize/kustomize/v5/commands/build.NewCmdBuild.func1
            0     0%     0%     40.12s 88.35%  sigs.k8s.io/kustomize/api/krusty.(*Kustomizer).Run
        0.51s  1.12%  1.12%     33.20s 73.11%  sigs.k8s.io/kustomize/api/resource.(*Resource).CurId
            0     0%  1.12%     26.95s 59.35%  sigs.k8s.io/kustomize/api/resmap.(*resWrangler).GetMatchingResourcesByCurrentId
        0.35s  0.77%  1.89%     26.95s 59.35%  sigs.k8s.io/kustomize/api/resmap.(*resWrangler).filteredById
        0.07s  0.15%  2.05%     25.53s 56.22%  sigs.k8s.io/kustomize/api/resmap.GetCurrentId
            0     0%  2.05%     21.68s 47.74%  sigs.k8s.io/kustomize/api/internal/target.(*KustTarget).MakeCustomizedResMap (inline)
            0     0%  2.05%     21.68s 47.74%  sigs.k8s.io/kustomize/api/internal/target.(*KustTarget).makeCustomizedResMap
        0.54s  1.19%  3.24%     19.75s 43.49%  sigs.k8s.io/kustomize/api/resource.(*Resource).GetGvk (inline)
            1s  2.20%  5.44%     19.21s 42.30%  sigs.k8s.io/kustomize/kyaml/resid.GvkFromNode
            0     0%  5.44%     18.42s 40.56%  sigs.k8s.io/kustomize/api/internal/builtins.(*SortOrderTransformerPlugin).Transform
            0     0%  5.44%     18.42s 40.56%  sigs.k8s.io/kustomize/api/krusty.(*Kustomizer).applySortOrder
            0     0%  5.44%     18.40s 40.52%  sigs.k8s.io/kustomize/api/internal/builtins.applyOrdering
        0.87s  1.92%  7.36%     16.55s 36.45%  sigs.k8s.io/kustomize/kyaml/yaml.visitMappingNodeFields
        2.51s  5.53% 12.88%     15.68s 34.53%  sigs.k8s.io/kustomize/kyaml/yaml.visitFieldsWhileTrue

after

    (pprof) top20 -cum
    Showing nodes accounting for 1.23s, 3.85% of 31.98s total
    Dropped 584 nodes (cum <= 0.16s)
    Showing top 20 nodes out of 184
        flat  flat%   sum%        cum   cum%
            0     0%     0%     27.41s 85.71%  github.com/spf13/cobra.(*Command).Execute
            0     0%     0%     27.41s 85.71%  github.com/spf13/cobra.(*Command).ExecuteC
            0     0%     0%     27.41s 85.71%  github.com/spf13/cobra.(*Command).execute
            0     0%     0%     27.41s 85.71%  main.main
            0     0%     0%     27.41s 85.71%  runtime.main
            0     0%     0%     27.41s 85.71%  sigs.k8s.io/kustomize/kustomize/v5/commands/build.NewCmdBuild.func1
            0     0%     0%     26.85s 83.96%  sigs.k8s.io/kustomize/api/krusty.(*Kustomizer).Run
            0     0%     0%     22.07s 69.01%  sigs.k8s.io/kustomize/api/internal/target.(*KustTarget).MakeCustomizedResMap (inline)
            0     0%     0%     22.07s 69.01%  sigs.k8s.io/kustomize/api/internal/target.(*KustTarget).makeCustomizedResMap
        0.38s  1.19%  1.19%     20.69s 64.70%  sigs.k8s.io/kustomize/api/resource.(*Resource).CurId
            0     0%  1.19%     13.64s 42.65%  sigs.k8s.io/kustomize/api/resmap.(*resWrangler).Append
            0     0%  1.19%     13.55s 42.37%  sigs.k8s.io/kustomize/api/resmap.(*resWrangler).GetMatchingResourcesByCurrentId (inline)
        0.12s  0.38%  1.56%     13.55s 42.37%  sigs.k8s.io/kustomize/api/resmap.(*resWrangler).filteredById
        0.01s 0.031%  1.59%     12.67s 39.62%  sigs.k8s.io/kustomize/api/resmap.GetCurrentId
        0.21s  0.66%  2.25%     12.49s 39.06%  sigs.k8s.io/kustomize/api/resource.(*Resource).GetGvk (inline)
        0.51s  1.59%  3.85%     12.28s 38.40%  sigs.k8s.io/kustomize/kyaml/resid.GvkFromNode
            0     0%  3.85%     11.52s 36.02%  sigs.k8s.io/kustomize/api/internal/target.(*KustTarget).IgnoreLocal
            0     0%  3.85%     10.53s 32.93%  sigs.k8s.io/kustomize/api/internal/target.(*KustTarget).AccumulateTarget
            0     0%  3.85%     10.53s 32.93%  sigs.k8s.io/kustomize/api/internal/target.(*KustTarget).accumulateResources
            0     0%  3.85%     10.53s 32.93%  sigs.k8s.io/kustomize/api/internal/target.(*KustTarget).accumulateTarget

koba1t commented:
/ok-to-test

koba1t commented on api/internal/builtins/SortOrderTransformer.go:
```diff
@@ -118,11 +109,13 @@ type legacyIDSorter struct {
 	// resids only stores the metadata of the object. This is an optimization as
 	// it's expensive to compute these again and again during ordering.
 	resids     []resid.ResId
+	resources []*resource.Resource
```
nit: Could you write what is the difference between this variable and the old variable `resids`.

chlunde replied:
Done, I used the comment suggested by @yanniszark

koba1t commented:
/lgtm

natasha41575 commented:
@yanniszark Do you have time to take a look at this one?

yanniszark commented on api/internal/builtins/SortOrderTransformer.go:
```diff
@@ -87,19 +87,10 @@ func (p *SortOrderTransformerPlugin) Transform(m resmap.ResMap) (err error) {
-		}
-	}
+func applyOrdering(m resmap.ResMap, resources []*resource.Resource) error {
 	m.Clear()
 	for _, r := range resources {
-		err = m.Append(r)
+		err := m.Append(r)
 		if err != nil {
 			return errors.WrapPrefixf(err, "SortOrderTransformer: Failed to append to resources")
 		}
```
I think there's no longer a need for a `applyOrdering` function.
The resources are already ordered so the function doesn't do much anymore.
Might as do that in the `Transform` function or rename this to `replaceResMapResources` or similar.

chlunde replied:
I will do a follow-up with another performance improvement in this file, so I hope we could merge this first and then do more cleanups there or later. I hope that's OK

yanniszark commented on api/internal/builtins/SortOrderTransformer.go:
```diff
@@ -118,11 +109,13 @@ type legacyIDSorter struct {
 	// resids only stores the metadata of the object. This is an optimization as
 	// it's expensive to compute these again and again during ordering.
 	resids     []resid.ResId
+	resources []*resource.Resource
```
I think it's worth adding a note here to explain why we do things this way and why it matters.
Basically, what happens is this:
- Initially, we sorted the metadata (ResId) of each object and then called `GetByCurrentId` on each to construct the  final list.
- The problem is that `GetByCurrentId` is inefficient and does a linear scan in a list every time we do that.
- So instead, we sort resources alongside the ResIds.

@natasha41575 I wonder if `GetByCurrentId` could use some better data structure in the future.

yanniszark reviewed (changes requested):
@chlunde thanks a lot for your perf PRs, they are greatly needed and will massively improve the user experience.
I left 2 comments for small changes but overall if this passes tests it should be good to go.

koba1t commented:
/triage accepted
/kind feature
/priority important-longterm

natasha41575 commented on api/internal/builtins/SortOrderTransformer.go:
```diff
@@ -87,19 +87,10 @@ func (p *SortOrderTransformerPlugin) Transform(m resmap.ResMap) (err error) {
 // applyOrdering takes resources (given in ResMap) and a desired ordering given
 // as a sequence of ResIds, and updates the ResMap's resources to match the
 // ordering.
-func applyOrdering(m resmap.ResMap, ordering []resid.ResId) error {
```
This file is generated code and should not be modified directly. 

Please update the corresponding file under `plugins` and use the `make` target that builds the plugins to generate this code.

chlunde replied:
thanks, fixed! `make generate-kustomize-builtin-plugins`

chlunde commented:
I believe I've addressed the comments, can you take another look? Thanks!

chlunde commented:
@natasha41575 or @yanniszark - can we get this merged? I'd like some input on how to structure the next PR #5427 but since it creates merge conflicts with this PR it would be better if this is merged first.

chlunde commented:
Hi, @koba1t can we get this merged? Thanks for taking a look :)

koba1t commented:
@natasha41575 
Could you recheck this PR?
I think you reviewed this PR and I want to hear your opinions.

chlunde commented:
@natasha41575 could you check this again or delegate to someone else?

chlunde commented:
> @chlunde thanks a lot for your perf PRs, they are greatly needed and will massively improve the user experience. I left 2 comments for small changes but overall if this passes tests it should be good to go.

@yanniszark I believe I've addressed your comments, could we get this merged now?

yanniszark commented:
Hi @chlunde and sorry for the late reply!
I don't have any merging capabilities unfortunately.
The current approvers are @varshaprasad96 @koba1t @knverey

koba1t commented:
HI @yanniszark 
If you think your commented point was fixed, Please type the `/lgtm` command.
https://github.com/kubernetes-sigs/kustomize/pull/5079#pullrequestreview-1336074940
That mark means you think this PR is ready to merge.

chlunde commented:
@koba1t could we get another reviewer here please? Soon 1 year birthday :cake:

yanniszark commented:
/lgtm

koba1t commented:
@chlunde 
Thanks, and I'm sorry for the delay in the review.
It looks like all test was passed; I think this PR is good to me!

/approve

koba1t commented:
/reopen

koba1t commented:
/lgtm

koba1t commented:
Sorry. Maybe I missed merging the last time.

chlunde commented:
@koba1t thanks! :+1:
